### ISSUE TITLE
Ignore some files by default?

### ISSUE THREAD

**Issue by gunwald** [syncthing/syncthing#837]
I would suggest to introduce a global ignore list, that has some default entries. For example:
   *.git/* # ignore git repros 
   *.sparkleshare/* # ignore git sparkleshare 
   *.~lock.* # ignore LibreOffice lock files
   *.owncloudsync* # ignore owncloud

It is annoying to add to each folder on each node every time all files you want to be ignored.

**Comment by Zillode** [comment]
I think we should reconsider this. We should ignore operating system files that are known to cause conflicts as soon as users start using Syncthing. My current list of such files consists of:

```
".DS_Store", // OSX (https://github.com/syncthing/syncthing/issues/826)
".Spotlight-V100", // OSX
"Thumbs.db", // Windows
"lost+found", // Linux (https://discourse.syncthing.net/t/un-ignore-subfolders-files-of-ignored-folder/1460/10 , https://github.com/syncthing/syncthing/issues/1090)
```
I suggest to list these as the default .stignore content.

**Comment by AudriusButkevicius** [comment]
`
".DS_Store",
"Thumbs.db",
"._*",
".Spotlight-V100",
".Trashes",
"desktop.ini",
`

@Zillode's list make sense,
I think we should have a default set of OS specific things (rather than .svn et al) @calmh ?

**Comment by facastagnini** [comment]
I build my exclusion list using the expressions relevant to my OS from here:

http://support.code42.com/CrashPlan/Latest/Troubleshooting/What_Is_Not_Backing_Up

I think is a comprehensive list.

**Comment by generalmanager** [comment]
I was going to propose my own list, but the CrashPlan one is way more comprehensive. Nice find @facastagnini !

**Comment by facastagnini** [comment]
You are welcome!

I started using Syncthing and Crashplan pretty much at the same time, so
when I was building a common ignore list for both tools I stumble upon that.

Since we are linking that list I feel it is appropriated to thanks them for
sharing that comprehensive ignore list, and give them some credit saying
that Crashplan is an awesome backup tool that complements my Syncthing
setup.

**Comment by flungo** [comment]
I am looking to implement this, if it hasn't already been started. Not sure if it would be best to have this hard coded into the binary, or if to create the global ignore list suggested populated by default with the system generated files. The global ignore list would probably be best located in the user config directory (alongside the config.xml).

I think as a minimum the list should include:

    .DS_Store
    .DS_Store?
    ._*
    .Spotlight-V100
    .Trashes
    .Trash-*
    Icon?
    ehthumbs.db
    desktop.ini
    Thumbs.db

I feel that the "comprehensive list" is a little too comprehensive, in the sense that, that list was written for full system backups.

**Comment by generalmanager** [comment]
@flungo It would be great if you found the time to do this. There are plans for global ignores in https://github.com/syncthing/syncthing/issues/1392. But as was already discussed over there those global ignores would not be synced to a master device for security reasons.

I would suggest a list of files in the main config file. The files in this list are added to the automatically created .stignore file in every folder at creation.

This would make it clear to everybody wondering why some files don't sync AND if someone really wants to change this it can either be done for all future folders on a device (by changing the config) or on a per-folder base.

Seems like the best solution to make it obvious and configurable.

**Comment by Zillode** [comment]
I agree with @flungo (wrt "comprehensive list") and @generalmanager for the implementation.
Feel free to initiate a PR @flungo

**Comment by flungo** [comment]
@generalmanager that definitely seems like the best idea actually. You're right that if it was global it would cause problems and potential confusion. I will see what I can come up with - doesn't sound too complicated a task though.

To be honest, with regards to the global exclude lists, I myself am quite happy to use symbolic/hard links (depending on if I want the global list to sync or not) and then using a `#include` statement in the folders `.stignore`.

If anyone sees any problems with the list I posted above then feel free to let me know what you think should be changed/added.

**Comment by generalmanager** [comment]
@flungo 

>To be honest, with regards to the global exclude lists, I myself am quite happy to use symbolic/hard links (depending on if I want the global list to sync or not) and then using a #include statement in the folders .stignore.

Yep, me too. But most users don't know what soft- or hardlinks are. And it doesn't prevent malicious nodes from introducing unwanted changes.

With regard to your list: I think it's a good base to work with, but where does "Icon?" come from? It's not in the CrashPlan list and might be broader than necessary.

**Comment by flungo** [comment]
@generalmanager 

So the `Icon?` is a file sometimes found on OSX machines. It is the file created if the user has set a custom folder icon in finder. The `?` is actually matching a `\r`. Could cause an issue if someone made a file called "Icon" or "Icons" I suppose...

**Comment by generalmanager** [comment]
@flungo Yeah, I read that \r was used instead of \n before OS X. We should definately test this. Would suck if Syncthing accidentally ignored Icon or Icon*.

**Comment by calmh** [comment]
`Icon\?` should match precisely that filename and not `Icons` etc.

**Comment by flungo** [comment]
@calmh Would that not match a literal `\` as opposed to `\r` (as in the return carriage symbol)

For config how does this sound:

    <configuration version="8">
        ...
        <defaults>
            <ignores>
                <ignore>.DS_Store</ignore>
                <ignore>.DS_Store?</ignore>
                <ignore>._*</ignore>
                <ignore>.Spotlight-V100</ignore>
                <ignore>.Trashes</ignore>
                ...
            </ignores>
        </defaults>
    </configuration>

The `<defaults>` section could be gotten rid of but I feel this would provide a nice expansion to being able to set other defaults when adding a new folder (such as a default versioning scheme). Realistically the `<ignores>` could be dropped too, but I feel that would start getting messy.

@generalmanager?

**Comment by calmh** [comment]
No, that matches a file called `Icon?`, with the question mark at the end. The backslash makes sure the question mark is interpreted as a literal question mark and not as a one character wildcard. (I'm not sure what the `\r` part is about?)

**Comment by flungo** [comment]
@calmh I was using it as a one character wildcard. On Mac OSX it uses an actual `\r` return carriage at the end of the file name. I believe in the terminal, it will be displayed as a `?` but the filename is `Icon\r`.

I will leave this one out until it can be confirmed.

**Comment by calmh** [comment]
Ah, yes, you're right, the file is actually called `Icon\r`. That file is the WTF that just keeps on giving... :) That's probably not possible to match with the current ignore language other than as `Icon?` which as you say is too wide.

**Comment by flungo** [comment]
`.DS_Store?` Is actually trying to match the same thing (with the ? being the `\r` character). Although I think its unlikely that someone will choose to make a file that's name matches `.DS_Store?`(the non question marked one is for on windows (possibly unix too) after you extract a zip that had a `.DS_Store\r` folder it turns it into just `.DS_Store`: Apple clearly thought it would be a great idea to distinguish their generated files by appending a seemingly invisible character to the end of the files. *sigh*

**Comment by flungo** [comment]
For reference/review, my current development branch: https://github.com/flungo/syncthing/tree/feature/issue/837

I will submit a PR when the feature is complete. If you see any problems though let me know so I can fix before making the PR.

**Comment by calmh** [comment]
The `.DS_Store` file is just that on Mac, no `\r` or anything. That particular madness is reserved for the Icon file, for historical reasons probably.

**Comment by flungo** [comment]
Okay so this feature is mostly complete: just needs testing.

One other possible thing to implement is whether during the upgrade, the new default ".stignore" should be added to existing folders that do not have a an ignore file already?

**Comment by generalmanager** [comment]
@flungo Wow, great work! I would personally prefer, if those were appended on upgrade, but I think we should leave it to @calmh to decide this.

**Comment by AudriusButkevicius** [comment]
This is great stuff.
The only idea I have is that perhaps these should be appended to the list after we've read from .stignore.
This way they would always be there, and you wouldn't need to go and change them everyone after you changed them in the cfg.

**Comment by flungo** [comment]
I was thinking about the pre/post append. If that were to be done how would it distinguish between purposefully modified ignore files and . It would be fine in the upgrade process to prepends the new defaults to the top and if a person modified it

But then yes, if you wanted to modify it slightly, it would drive you crazy going through every folder you had synced and changing this. It could be done similarly to how wordpress manages the .htaccess file my putting a marker (such as `\\---BEGIN SYNCTHING DEFAULTS---\\` and `\\---END SYNCTHING DEFAULTS---\\` around the generated section, but then if this is removed after creation/upgrade then it will no longer update this during the folder checks.

One other potential idea I had was if the config file stored the ignore list (possibly multiple of them identified by an attribute) which could be referenced from the syncthing `.stignore` with some new syntax (such as `@include defaults`). In one way this would be great as it could centralise the management of these ignore lists and allow a user to define custom ones (like a git one, or a owncloud one - giving the OP a solution that would work for their needs) that can be added to any `.syncignore`. So to reiterate for clarity `#include` would include from the local directory, whereas `@include` (or similar) would do a lookup in the config for an ignore list.

Let me know what you guys are thinking. I will try and do some testing in the meanwhile.

**As a side note:** What have you other developers/collaborators found is the best way to test? I will see if there any relevant unit tests that need to be added then probably set up a docker image that I can create and dispose of quickly to test various start conditions and check that an upgrade works, etc.

**Comment by AudriusButkevicius** [comment]
There are a bunch of tests which test ignore files, as well as integration tests to make sure stuff still works.
The whole `#include` vs `@include` is more or less redundant I think, because you can already achieve what you are talking about with file includes. 

The only benefit of @include over #include is that you can reuse it in multiple folders, but it still involves a manual step of adding `@include defaults`, which I guess what this ticket was trying to avoid.

I don't mind having `@include` and `#include`, but I'd prefer to avoid having people to add `@include defaults`.

Perhaps we can have `@include` and `#include`, but then always silently include `@include defaults`?

Also, next feature I see coming is synced @include's which is #1408 and #1392

**Comment by flungo** [comment]
With regards to the `@include` suggested - I would suggest automatically and silently adding it to the start of every file (during upgrade and when creating a new folder) with a good comment above about its purpose and where it can be edited and a reference to the [documentation](https://github.com/syncthing/syncthing/wiki/Ignoring-Files). Similar to what I have already implemented (here)[https://github.com/flungo/syncthing/blob/0cb641775ffa5bb20341429f62c0ff589ab19d9f/internal/config/config.go#L108-L114]

However if the user chooses to remove it the `@include default` line, then we assume they know what they are doing and it should stay removed.

**Comment by AudriusButkevicius** [comment]
Makes sense.

**Comment by generalmanager** [comment]
@flungo I agree, seems like the cleanest and most logical way to me.

**Comment by flungo** [comment]
Just going to post these links here for reference:

https://github.com/github/gitignore/blob/master/Global/Linux.gitignore
https://github.com/github/gitignore/blob/master/Global/Windows.gitignore
https://github.com/github/gitignore/blob/master/Global/OSX.gitignore

Not all relevant but its pretty obvious which ones we do and don't want from each list.

**Comment by flungo** [comment]
Hectic week with work, hoping to have this finished by next weekend at the latest. Sooner hopefully though.

**Comment by counterbeing** [comment]
Right on! This seems very much needed. I think a `~/.stignore` would be perfect. Just file separated by lines containing patterns. Simple and to the point, just like `.gitignore`. 

Thanks! Just got into SyncThing, and it is awesome!

**Comment by sysfu** [comment]
Any status update on implementation of the global ignore feature? I just started syncing between OS X and Windows and the hidden file litter is driving me crazy.

**Comment by AudriusButkevicius** [comment]
Read the manual/docs, there is already an include directive which allows doing that.

**Comment by sysfu** [comment]
Look, I'm all too familiar with the tedious and error prone manual method of using include statements.

The issue the original poster raised was that of shipping a global ignore list "that has some default entries".

The list suggested by @flungo below is a fine one. So how about it? How about shipping this global list and enabling it by default with new installations? And also exposing it in the web interface for editing?

The people with text editing fetishes will still be quite free to edit/include/link all the myriad of .stignore files they want and religiously check them into their version control systems.

```
.DS_Store
.DS_Store?
._*
.Spotlight-V100
.Trashes
.Trash-*
Icon?
ehthumbs.db
desktop.ini
lost+found
Thumbs.db
```

**Comment by AudriusButkevicius** [comment]
It's already editable from the web ui?
Have you even tried syncthing?

**Comment by generalmanager** [comment]
@sysfu Agreed, I'd like to see this make it's way into production. I have spent countless hours copying selfmade ignore lists on any mixed infrastructure syncthing setup I can think of. The average enduser does not have the kind of knowledge necessary to fix something like this.
It's one of those many nice details successful products like Dropbox get right:
Design for the stupidest and most technologically challenged user you can think of. Then replace him with a 3 year old. Now with a monkey. If your product still works, it's great.
If not, it's never gonna have the kind of adoption necessary to change the lives of the average joe.

Sorry for the big words, but having a problem-free experience right from the start is incredibly important to make a tool usable for most people.

**Comment by sysfu** [comment]
A global ignore list/patterns feature with sane defaults already exists which is editable from the web interface, ORLY?

Let's take a look, shall we...

'Actions' menu => Settings   Anything like 'Edit Global ignore pattens'? Nope.

Ok, let's try again:
'Actions' menu => 'Advanced Settings' => 'Options' section. Anything like edit 'Edit Global ignore pattens'? Nope.

And again:
'Actions' menu => 'Advanced Settings' => 'Options' section. Anything like 'Edit Global ignore pattens'? Nope.

Now of course a user can select a folder, click the 'Edit' button, and then the 'Ignore Patterns' button, but this is neither global nor does it include any sane defaults.

Fail.

**Comment by simplypeachy** [comment]
@Zillode On Windows systems you may want to ignore some of the system directories/files that can sometimes be found at the root of drives; some users may be sharing out an entire drive. Although not necessarily the system drive - I expect most people know this isn't a good idea. (Or would they?)

* Files:
  * pagefile.sys (could be found on any fixed drive)

* Directories:
  * $Recycle.Bin
  * System Volume Information

hiberfil.sys and swapfile.sys can only appear outside of the system drive's root by hacking Windows, so I don't think ignores for them are necessary.

On a non-OS related note - I ignore browser "partial download" files and Zip temporary files as a result of past experiences. If these patterns are of use:
Firefox and IE:
`\.part(ial)?$`
Chrome:
`\.crdownload$`
Info-ZIP:
`^zia[0-9]+$`
WinZip:
`^_Za[0-9]+$`

**Comment by uok** [comment]
there are also swapfile.sys and hiberfil.sys

**Comment by jonathancross** [comment]
Hi @Zillode and others, what is the status on this?  Is there anything blocking in https://github.com/Zillode/syncthing/commit/d15af03e88bb55d636facba5581f5c622baa474a ?

**Comment by calmh** [comment]
Why would I not want to sync my thumbnail caches if I'm syncing the folder containing the images in question?

(The deeper question I\m hinting at here is "why do you think your ignore patterns are the ones I want?")

**Comment by jhasse** [comment]
> Why would I not want to sync my thumbnail caches if I'm syncing the folder containing the images in question?

Because you might be running two versions of Windows that are use an incompatible Thumbs.db format and now you're getting conflicts for a file, which might just as well be generated on the fly.

Imagine someone adding an image and having Explorer open. Syncthing synces the image and then Thumbs.db. Another Windows node might receive the image first and its Explorer instance also regenerates the Thumbs.db file -> conflict.

**Comment by canton7** [comment]
> (The deeper question I'm hinting at here is "why do you think your ignore patterns are the ones I want?")

It's a valid point. However I think that there is a small, limited subset of ignores which we can say with 99.99% confidence that users should ignore. E.g. NTUSER.dat, the recycle bin, System Volume Information (for Windows).

**Comment by calmh** [comment]
I agree for those. And I'm not saying there is *never* cause to ignore thumbs.db, just that if it's not clearly the right thing to do I don't think we should do it automagically for the user. So I think the set of files to auto ignore is a bit smaller than what's suggested in a couple of places. For example the list of things git autoignores (links a fair bit up) is super inappropriate for us.

**Comment by jhasse** [comment]
> So I think the set of files to auto ignore is a bit smaller than what's suggested in a couple of places. For example the list of things git autoignores (links a fair bit up) is super inappropriate for us.

I agree.

But what about my Thumbs.db conflict example? Don't you agree that 99% of users would want to ignore that file? (Also keep in mind that Thumbs.db isn't invisible on Linux, for that reason alone I always want to ignore it)

**Comment by canton7** [comment]
Agreed. I was just making sure we didn't discount the idea entirely :smile: 

(On the thumbs.db point, I agree with @jhasse that it should be ignored, but we're heading off into bikeshedding territory...)

**Comment by calmh** [comment]
Yeah just making conversation on issues that popped on my notification list... We're going to bikeshed the shit out of this when it becomes a PR, that's for sure. 🚲 🏠🖌

**Comment by Ferroin** [comment]
Having a default set and then the ability to add to that list on a per-device basis would be nice.

I would assume the basic list should look something like:
```
# From Windows
$RECYCLE.BIN
System Volume Information
$WINDOWS.~BT
pagefile.sys
desktop.ini

# From OS X
Icon?
.Spotlight-V100
.Trashes
.DS_Store
.fseventsd

# From Linux
lost+found
.gvfs
.local/share/trash
.Trash-*

# Generic stuff
._*
```

$RECYCLE.BIN (case insensitive, since newer Windows versions use $Recycle.Bin) is the per-drive recycling bin for Windows
$WINDOWS.~BT is a temporary directory used for OS upgrades on Windows (and is often _huge_ if it exists).  I've only ever seen this on the C: drive, but I think it shows up on a different drive if there's not enough room on C:.

.fseventsd is something I see a lot on removable media.  I'm not 100% certain what OS it's from, but it looks like ancillary volume metadata (which makes me think it's probably OS X), so it probably shouldn't be synced.

.gvfs is found in most people's home directories on Linux when they use most desktop environments other than KDE, and shouldn't be scanned because it covers dynamic access to access transiently connected and non-filesystem items (network filesystems, archives, MTP devices, etc)
.local/share/trash is the per-user trash-can for XDG compliant desktop environments on Linux (which is 90% of what people who use X use on Linux).

Regardless of what platform we're on, we should use the same default excludes for consistency (and because people do share storage devices between systems).

**Comment by CONSULitAS** [comment]
Hi,

i am just having the problem of having an error while syncing with the OS X Icon file (`filename is invalid`)!

After thinking about the Problem that @flungo mentioned, that Icon? ignores the rather common `Icons` i would suggest the following:

```
!Icon[0-9a-zA-Z]
(?d)Icon?
```

Regards

**Comment by NatoBoram** [comment]
There's already a list of sane `.gitignore` : https://www.gitignore.io/api/linux,macos,windows

If Syncthing could simply use the .gitignore format and extend it for their needs, I think it would've been a better solution.

Then, to automate adding the ignore file on each machine and each folders, one could simply `curl https://www.gitignore.io/api/linux,macos,windows -o .stignore` each folders.

**Comment by AudriusButkevicius** [comment]
We can't use gitignore because our format is already different, plus it has control sequences that control what happens if the file is detected ((?d)/(?i)) etc.

**Comment by NatoBoram** [comment]
Then Syncthing should add a feature to migrate the well-known `.gitignore` file into their custom `.stignore`. It's the one inventing their own format, they should be responsible for creating an easy way to use it.

**Comment by AudriusButkevicius** [comment]
It shouldn't, because there is disagreement in what the format is able to describe, so if we were to follow what you are suggesting, we'd have to remove features that people currently rely on. 

The .gitignore ship has sailed many moons ago, so people should stop bringing this up, as this is going nowhere.

I think people should stop bringing up

**Comment by calmh** [comment]
Git itself is also a super niche tool used by a tiny clique of people in the world. I realize it’s a vocal minority here given the site we’re on, but that doesn’t really matter in the grander scheme of Syncthing users generally. So what Audrius says, let’s stop talking about gitignores.

**Comment by hyiltiz** [comment]
So far, trying to sync a folder which somehow contained (not directly) a version control repository (e.g. `.git`) corrupts the `.git/*` objects **unpredictably** --- it doesn't happen all the time and doesn't corrupt all the repos; just happens sometimes for some repos under some weird race(?) conditions. We can argue if `.git` is a *niche* tool but consider this: the number that really matters here is the conditional probability of a syncthing user also using `git` aka `p(uses git | uses syncthing)`. I bet this number is fairly high, meaning we should not ignore a large portion of our current user base.

There is a much stronger point here about usability by default; *a backup software should never corrupt what is already there*. Syncthing by default *can* and *will* corrupt `.git` projects at some point. I have a few projects of mine corrupted and had to `rm -rf .git` and start afresh again. 

The solution, however, breaks nothing and is simple; just include default ignore lists. Yes, we cannot have an exhaustive list of everything one would like to ignore, but that shouldn't stop us from trying our best.

**Comment by beerse** [comment]
Wild idea on extending/updating the .stignore feature:
First: Use the file '.stignore' (the current filename) and add to that the directory '.stignore.d' and all files (recursive) in it.
Reason: There everyone can ad there <toolname>  and/or <osname> with dedicated ignore patterns for that area.

Second: Read the .stignore and .stignore.d (in this order) from the software installation location (default: file containing ".stignore"), the 'standard map location' and the root of the current tree-to-sync. 

Major change: do sync the .stignore file and .stignore.d tree in the current tree-to-sync. If you donnot want to do so, Just add "/.stignore" to one of your .stignore files.
To keep up with the current configuration: Add "/.stignore" to the .stignore in the installation directory. The I can put "!./stignore in the .stignore file in the three and it is synced to my wishes.
